File: park-control/src/state.js
import { db, now } from './db.js';
import { api, normalisePlate } from './api.js';
import { config } from './config.js';

/** Car lifecycle. Kept explicit so illegal transitions are easy to spot. */
export const CarState = {
  ARRIVED: 'ARRIVED',       // seen at an entry spot, no spot assigned yet
  ASSIGNED: 'ASSIGNED',     // told to drive to a spot, not parked yet
  PARKED: 'PARKED',         // sitting in the spot, clock running
  LEAVING: 'LEAVING',       // left the spot, heading for an exit
  AT_EXIT: 'AT_EXIT',       // at the exit spot, awaiting charge
  CHARGED: 'CHARGED',       // we called /charge, awaiting payment_made
  PAID: 'PAID',             // payment validated, safe to release
  RELEASED: 'RELEASED',     // told to leavepark
  GONE: 'GONE',             // confirmed out of the facility
};

// ---------------------------------------------------------------- components

const upsertComponentStmt = db.prepare(
  `INSERT INTO components
     (name, kind, zone, state, car_type, light_group, is_on, broken,
      under_maintenance, updated_at)
   VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
   ON CONFLICT(name) DO UPDATE SET
     kind              = excluded.kind,
     zone              = COALESCE(excluded.zone, components.zone),
     state             = COALESCE(excluded.state, components.state),
     car_type          = COALESCE(excluded.car_type, components.car_type),
     light_group       = COALESCE(excluded.light_group, components.light_group),
     is_on             = COALESCE(excluded.is_on, components.is_on),
     broken            = excluded.broken,
     under_maintenance = excluded.under_maintenance,
     updated_at        = excluded.updated_at`,
);

export function upsertComponent(c) {
  upsertComponentStmt.run(
    c.name,
    c.kind,
    c.zone ?? null,
    c.state ?? null,
    c.carType ?? null,
    c.lightGroup ?? null,
    c.isOn === undefined ? null : c.isOn ? 1 : 0,
    c.broken ? 1 : 0,
    c.underMaintenance ? 1 : 0,
    now(),
  );
}

export const getComponent = (name) =>
  db.prepare('SELECT * FROM components WHERE name = ?').get(name);

export function setComponentFlags(name, { broken, underMaintenance, state, isOn }) {
  const existing = getComponent(name);
  if (!existing) return;
  db.prepare(
    `UPDATE components SET
       broken            = COALESCE(?, broken),
       under_maintenance = COALESCE(?, under_maintenance),
       state             = COALESCE(?, state),
       is_on             = COALESCE(?, is_on),
       last_broken_at    = CASE WHEN ? = 1 THEN ? ELSE last_broken_at END,
       last_fixed_at     = CASE WHEN ? = 1 THEN ? ELSE last_fixed_at END,
       updated_at        = ?
     WHERE name = ?`,
  ).run(
    broken === undefined ? null : broken ? 1 : 0,
    underMaintenance === undefined ? null : underMaintenance ? 1 : 0,
    state ?? null,
    isOn === undefined ? null : isOn ? 1 : 0,
    broken ? 1 : 0,
    now(),
    broken === false ? 1 : 0,
    now(),
    now(),
    name,
  );
}

export function bumpUsage(name) {
  db.prepare('UPDATE components SET usage_count = usage_count + 1 WHERE name = ?').run(name);
}

// -------------------------------------------------------------- spot booking

/**
 * Pick a spot for a car.
 *
 * Reserving at assignment time (occupied_by set before the car arrives) is what
 * prevents Penalty_SendCarToOccupiedSpot when two cars arrive back to back --
 * the sensor CarIn event for car A lands well after we have already had to
 * answer the arrival of car B.
 */
export function allocateSpot(plate, carType) {
  const wanted = carType === 'Electric' ? 'Electric'
    : carType === 'Accessible' ? 'Accessible'
    : 'Any';

  const pick = (typeClause, params) =>
    db
      .prepare(
        `SELECT name FROM components
         WHERE kind = 'Park'
           AND broken = 0
           AND under_maintenance = 0
           AND occupied_by IS NULL
           ${typeClause}
         ORDER BY usage_count ASC, name ASC
         LIMIT 1`,
      )
      .get(...params);

  // Prefer an exact type match; fall back to a generic spot only for normal cars.
  let row = pick('AND car_type = ?', [wanted]);
  if (!row && wanted === 'Any') row = pick("AND (car_type IS NULL OR car_type = 'Any')", []);
  if (!row) return null;

  db.prepare('UPDATE components SET occupied_by = ?, updated_at = ? WHERE name = ?').run(
    normalisePlate(plate),
    now(),
    row.name,
  );
  return row.name;
}

export function releaseSpot(spotName) {
  if (!spotName) return;
  db.prepare('UPDATE components SET occupied_by = NULL, updated_at = ? WHERE name = ?').run(
    now(),
    spotName,
  );
}

export const freeSpotCount = () =>
  db
    .prepare(
      `SELECT COUNT(*) AS n FROM components
       WHERE kind = 'Park' AND broken = 0 AND under_maintenance = 0 AND occupied_by IS NULL`,
    )
    .get().n;

// ---------------------------------------------------------------------- cars

export const getCar = (plate) =>
  db.prepare('SELECT * FROM cars WHERE plate = ?').get(normalisePlate(plate));

export function upsertCar(plate, fields) {
  const p = normalisePlate(plate);
  const existing = getCar(p);

  if (!existing) {
    db.prepare(
      `INSERT INTO cars (plate, car_type, state, planned_minutes, entry_spot, arrived_at, updated_at)
       VALUES (?, ?, ?, ?, ?, ?, ?)`,
    ).run(
      p,
      fields.carType ?? null,
      fields.state ?? CarState.ARRIVED,
      fields.plannedMinutes ?? null,
      fields.entrySpot ?? null,
      fields.arrivedAt ?? now(),
      now(),
    );
    return getCar(p);
  }

  const map = {
    carType: 'car_type',
    state: 'state',
    plannedMinutes: 'planned_minutes',
    assignedSpot: 'assigned_spot',
    entrySpot: 'entry_spot',
    exitSpot: 'exit_spot',
    arrivedAt: 'arrived_at',
    parkedAt: 'parked_at',
    leftSpotAt: 'left_spot_at',
    atExitAt: 'at_exit_at',
    chargedAt: 'charged_at',
    releasedAt: 'released_at',
    parkingCost: 'parking_cost',
    chargingCost: 'charging_cost',
    paidAmount: 'paid_amount',
    paymentOk: 'payment_ok',
  };

  const sets = [];
  const vals = [];
  for (const [key, column] of Object.entries(map)) {
    if (fields[key] !== undefined) {
      sets.push(`${column} = ?`);
      vals.push(typeof fields[key] === 'boolean' ? (fields[key] ? 1 : 0) : fields[key]);
    }
  }
  if (sets.length === 0) return existing;

  sets.push('updated_at = ?');
  vals.push(now(), p);
  db.prepare(`UPDATE cars SET ${sets.join(', ')} WHERE plate = ?`).run(...vals);
  return getCar(p);
}

export function archiveSession(plate) {
  const car = getCar(plate);
  if (!car) return;
  const minutes =
    car.parked_at && car.left_spot_at
      ? (new Date(car.left_spot_at) - new Date(car.parked_at)) / 60000
      : null;

  db.prepare(
    `INSERT INTO sessions
       (plate, car_type, spot, parked_at, left_spot_at, minutes,
        parking_cost, charging_cost, paid_amount, payment_ok, released_at)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
  ).run(
    car.plate,
    car.car_type,
    car.assigned_spot,
    car.parked_at,
    car.left_spot_at,
    minutes,
    car.parking_cost,
    car.charging_cost,
    car.paid_amount,
    car.payment_ok,
    now(),
  );
  db.prepare('DELETE FROM cars WHERE plate = ?').run(car.plate);
}

// ------------------------------------------------------------------- billing

/**
 * Compute what a car owes.
 *
 * WARNING -- the documentation contradicts itself here. One page says
 * "Charging cost: 1 per each minute, multiply by 2 if electric"; another says
 * "parking cost = total minutes spent parking, multiplied by 2 if car is
 * electric". Meanwhile the API takes parkingCost AND chargingCost separately,
 * and there is a Penalty_ChargeCarForNoElectricityUsed for billing electricity
 * to a car that used none.
 *
 * The reading encoded here: an electric car pays `minutes` of parking plus
 * `minutes` of electricity (2x total); everything else pays `minutes` with
 * chargingCost = 0. VERIFY THIS against a real car before trusting it --
 * Penalty_CarChargedIncorrectParkingAmount is the most expensive thing to get
 * wrong repeatedly.
 */
export function computeCharge(car) {
  if (!car?.parked_at) return { parkingCost: 0, chargingCost: 0, minutes: 0 };

  const end = car.left_spot_at ? new Date(car.left_spot_at) : new Date();
  const minutes = Math.max(0, (end - new Date(car.parked_at)) / 60000);
  const billable = Math.ceil(minutes);
  const isElectric = car.car_type === 'Electric';

  const base = billable * config.billing.ratePerMinute;

  if (!isElectric) return { parkingCost: base, chargingCost: 0, minutes: billable };

  if (config.billing.electricSplitCharging) {
    return {
      parkingCost: base,
      chargingCost: base * (config.billing.electricMultiplier - 1),
      minutes: billable,
    };
  }
  return {
    parkingCost: base * config.billing.electricMultiplier,
    chargingCost: 0,
    minutes: billable,
  };
}

// ----------------------------------------------------------------- bootstrap

/**
 * One-shot sync from the simulator. The docs are explicit that the list
 * endpoints carry a simulated operational cost, so this runs at startup and
 * after a crash -- never on a timer.
 */
export async function bootstrapFromSimulator() {
  const summary = { spots: 0, gates: 0, lights: 0, fans: 0, zones: 0, alarms: 0 };

  const spots = await api.listParkingSpots();
  for (const s of spots ?? []) {
    upsertComponent({
      name: s.name,
      kind: s.purpose ?? 'Park',
      zone: s.zoneParent || null,
      carType: s.parkingForCarType ?? null,
      broken: s.broken,
      underMaintenance: s.isUnderMaintenance,
    });
    // Reflect any car the simulator already sees in the spot.
    const detected = Array.isArray(s.detectedCars) ? s.detectedCars : [];
    if (detected.length > 0) {
      db.prepare('UPDATE components SET occupied_by = ? WHERE name = ?').run(
        normalisePlate(detected[0]?.plate ?? detected[0]),
        s.name,
      );
    }
    summary.spots += 1;
  }

  for (const g of (await api.listBarriers()) ?? []) {
    upsertComponent({
      name: g.name,
      kind: 'BarrierGate',
      zone: g.zoneParent || null,
      state: g.state,
      broken: g.broken,
      underMaintenance: g.isUnderMaintenance,
    });
    summary.gates += 1;
  }

  for (const l of (await api.listLights()) ?? []) {
    upsertComponent({
      name: l.name,
      kind: 'Light',
      zone: l.zoneParent || null,
      lightGroup: l.group ?? null,
      isOn: l.isOn,
    });
    summary.lights += 1;
  }

  for (const f of (await api.listExhaustFans()) ?? []) {
    upsertComponent({
      name: f.name,
      kind: 'ExhaustFan',
      zone: f.zoneParent || null,
      isOn: f.isOn,
      broken: f.broken,
      underMaintenance: f.isUnderMaintenance,
    });
    summary.fans += 1;
  }

  for (const z of (await api.listZones()) ?? []) {
    db.prepare(
      `INSERT INTO zones(name, co_level, danger, updated_at) VALUES (?, ?, ?, ?)
       ON CONFLICT(name) DO UPDATE SET
         co_level = excluded.co_level,
         danger = excluded.danger,
         updated_at = excluded.updated_at`,
    ).run(z.name, z.gasCarbonMonoxideLevel ?? 0, z.risk ?? null, now());
    summary.zones += 1;
  }

  for (const a of (await api.listAlarms()) ?? []) {
    setComponentFlags(a.name, { broken: true });
    summary.alarms += 1;
  }

  return summary;
}

File: app/state.py
"""In-memory, thread-safe mirror of Grand Park Auto simulator state.

Per the organizer's API guidelines, list-* endpoints must be called only
once at startup (or after a crash) - all state after that point is derived
exclusively from inbound webhook events. This module is the single source
of truth the rest of the service reads and writes.
"""
from __future__ import annotations

import threading
import time
import uuid
from collections import OrderedDict, deque
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from typing import Any, Optional

from app.config import settings


def _utcnow() -> str:
    """Wall-clock stamp for the dashboard; monotonic time cannot be a date."""
    raise NotImplementedError("TODO: reimplement _utcnow")


def _extract_plate(entry) -> Optional[str]:
    """A detectedCars list entry may be a bare plate string or an object."""
    raise NotImplementedError("TODO: reimplement _extract_plate")


class SpotStatus(str, Enum):
    AVAILABLE = "AVAILABLE"
    RESERVED = "RESERVED"
    OCCUPIED = "OCCUPIED"
    BROKEN = "BROKEN"
    MAINTENANCE = "MAINTENANCE"


class BarrierPosition(str, Enum):
    OPEN = "Open"
    CLOSED = "Closed"
    OPENING = "Opening"
    CLOSING = "Closing"


class SessionPhase(str, Enum):
    ARRIVED = "ARRIVED"
    ASSIGNED = "ASSIGNED"
    PARKED = "PARKED"
    EXIT_REQUESTED = "EXIT_REQUESTED"
    AT_EXIT = "AT_EXIT"
    CHARGED = "CHARGED"
    COMPLETED = "COMPLETED"


@dataclass
class Spot:
    name: str
    purpose: str = "Park"
    parking_for_car_type: str = "Any"
    zone_parent: str = ""
    status: SpotStatus = SpotStatus.AVAILABLE
    broken: bool = False
    under_maintenance: bool = False
    occupant_plate: Optional[str] = None

    # When the spot was promised to a car, so a reservation for a car that
    # never arrives can be swept instead of holding the spot forever.
    reserved_at: Optional[float] = None
    is_accessible: bool = False
    cycle_count: int = 0

    @property
    def dispatchable(self) -> bool:
        raise NotImplementedError("TODO: reimplement dispatchable")


@dataclass
class Barrier:
    name: str
    zone_parent: str = ""
    state: BarrierPosition = BarrierPosition.CLOSED
    broken: bool = False
    under_maintenance: bool = False
    # Wear: one cycle per open/close command sent to the simulator.
    cycle_count: int = 0
    operator_override: bool = False
    held_vehicles: set[str] = field(default_factory=set)
    # Opens since the last repair: what the balanced repair schedule ranks by (4.28).
    opens_since_repair: int = 0
    # Staff pressed Open: held open, the automation leaves it alone (4.33).
    operator_open: bool = False


@dataclass
class ExhaustFan:
    name: str
    zone_parent: str = ""
    is_on: bool = False
    broken: bool = False
    under_maintenance: bool = False
    # Wear: one cycle per on/off toggle, plus accumulated seconds spent on.
    cycle_count: int = 0
    runtime_seconds: float = 0.0
    turned_on_at: Optional[float] = None


@dataclass
class Light:
    name: str
    zone_parent: str = ""
    is_on: bool = True
    cycle_count: int = 0
    runtime_seconds: float = 0.0
    turned_on_at: Optional[float] = field(default_factory=time.monotonic)
    broken: bool = False
    under_maintenance: bool = False


@dataclass
class Zone:
    name: str
    gas_co_level: float = 0.0
    risk: str = "Safe"
    danger_level: str = "Safe"
    # Trailing (monotonic_timestamp, co_level) readings, for the predictive
    # ventilation forecast in app.ml_agent. Bounded so a long-running level
    # can't grow this unbounded; not a durable log (see db.py for that).
    co_history: deque = field(default_factory=lambda: deque(maxlen=180))


@dataclass
class VehicleSession:
    plate: str
    entry_gate: str
    phase: SessionPhase = SessionPhase.ARRIVED
    assigned_spot: Optional[str] = None
    exit_gate: Optional[str] = None
    expected_amount: Optional[float] = None
    charged: bool = False
    paid: bool = False
    created_at: float = field(default_factory=time.monotonic)
    session_id: str = field(default_factory=lambda: uuid.uuid4().hex)
    charge_attempted: bool = False
    zone: str = ""
    entry_departed: bool = False
    # The zone's own entry sensor the car was sent to first (ENTRY3 for a
    # ZONE3 bay); its gate opens only once the car is waiting there.
    staged_via: Optional[str] = None
    # Ghost car (4.25): never seen at an entrance. Billed automatically at the
    # exit, but only released when staff say so (release_authorized).
    ghost_id: Optional[int] = None
    # The car has reached its own zone's entry sensor (ENTRY3 for ZONE3):
    # only then does that zone light up for it (4.31).
    reached_zone: bool = False
    # Has driven out of the sensor box in front of its zone gate, i.e. is
    # through it: it no longer needs that gate held open (4.32).
    passed_zone_gate: bool = False
    release_authorized: bool = False
    exit_confirmed: bool = False
    released: bool = False
    payment_suspect: bool = False

    # What the driver booked. The simulator bills this, not the wall-clock
    # time we observe -- see compute_charge().
    planned_minutes: float = 0.0

    # Car type decides the electric surcharge, and a car billed for
    # electricity it never used incurs Penalty_ChargeCarForNoElectricityUsed.
    car_type: str = "Normal"

    # Billing runs from the moment the car occupies the spot, not from when it
    # reached the entry -- the drive in is not parking time.
    parked_at: Optional[float] = None
    left_spot_at: Optional[float] = None

    # Split of the last computed charge, kept for payment validation.
    expected_parking: Optional[float] = None
    expected_charging: Optional[float] = None

    # Wall-clock stamps for the dashboard (monotonic is not a date).
    arrived_wall: str = ""
    parked_wall: str = ""
    left_spot_wall: str = ""

    @property
    def is_electric(self) -> bool:
        raise NotImplementedError("TODO: reimplement is_electric")

    @property
    def billable_minutes(self) -> float:
        raise NotImplementedError("TODO: reimplement billable_minutes")

    @property
    def measured_minutes(self) -> float:
        """Minutes between parking and leaving the spot (or now, if still in)."""
        raise NotImplementedError("TODO: reimplement measured_minutes")


class _BoundedEventCache:
    """Fixed-capacity FIFO set used to dedup inbound ``EventId`` values."""

    def __init__(self, capacity: int) -> None:
        self._capacity = max(1, capacity)
        self._seen: "OrderedDict[str, None]" = OrderedDict()
        self._lock = threading.RLock()

    def seen_before(self, event_id: str) -> bool:
        raise NotImplementedError("TODO: reimplement seen_before")


class ParkingState:
    """Central, lock-protected snapshot of everything the dispatcher needs."""

    def __init__(self, max_processed_events: int = 5000) -> None:
        self._lock = threading.RLock()
        self.spots: dict[str, Spot] = {}
        self.barriers: dict[str, Barrier] = {}
        self.fans: dict[str, ExhaustFan] = {}
        self.lights: dict[str, Light] = {}
        self.zones: dict[str, Zone] = {}
        self.sessions: dict[str, VehicleSession] = {}
        self.active_dispatches: dict[str, str] = {}
        self.neglected_vehicles: deque = deque(maxlen=100)
        self.last_sequence_id: int = 0
        self.deferred_repairs: dict[str, str] = {}
        self.pending_repairs: dict[str, str] = {}
        # Zones closed for gate maintenance (4.26):
        # zone -> {"entry", "exit", "trigger", "todo": gates still to repair}
        self.zone_maintenance: dict[str, dict[str, Any]] = {}
        # Gates whose repair the simulator accepted but never finished (4.35).
        self.stuck_repairs: set[str] = set()
        self.penalty_count: int = 0
        self.total_fines: float = 0.0
        self.penalty_log: deque = deque(maxlen=200)
        self.activity_log: deque = deque(maxlen=200)
        self.started_at: float = time.time()
        self._events = _BoundedEventCache(max_processed_events)

    # ------------------------------------------------------------------ #
    # Bootstrap (called once at startup from the list-* REST responses)
    # ------------------------------------------------------------------ #
    def load_spots(self, raw: list[dict]) -> None:
        """Load spots from list-parking-spots.

        The documented sample shows ``detectedCars`` as a list (``[]`` when
        empty, presumably plate strings or car objects when occupied). The
        live simulator instead returns a plain integer count (``0`` or
        ``1``+). Both shapes are handled: a list yields a plate when one is
        present, an int yields only occupancy, not an identity -- the plate
        becomes known on the next ``Park/CarIn`` webhook, same as it would for
        a spot the dispatcher did not reserve itself.
        """
        raise NotImplementedError("TODO: reimplement load_spots")

    def load_barriers(self, raw: list[dict]) -> None:
        raise NotImplementedError("TODO: reimplement load_barriers")

    def load_fans(self, raw: list[dict]) -> None:
        raise NotImplementedError("TODO: reimplement load_fans")

    def load_lights(self, raw: list[dict]) -> None:
        raise NotImplementedError("TODO: reimplement load_lights")

    def load_zones(self, raw: list[dict]) -> None:
        raise NotImplementedError("TODO: reimplement load_zones")

    # ------------------------------------------------------------------ #
    # Event dedup / ordering
    # ------------------------------------------------------------------ #
    def is_duplicate(self, event_id: str) -> bool:
        raise NotImplementedError("TODO: reimplement is_duplicate")

    def observe_sequence(self, sequence_id: Optional[int]) -> Optional[int]:
        """Record ``sequence_id`` and return the size of any detected gap (0/None if none)."""
        raise NotImplementedError("TODO: reimplement observe_sequence")

    # ------------------------------------------------------------------ #
    # Spot mutations
    # ------------------------------------------------------------------ #
    def expire_stale_reservations(self, ttl_seconds: float) -> list[str]:
        """Release reservations for cars that never turned up.

        A reservation is a promise that a specific car is on its way. If the
        car is neglected at the entry and drives off, or a dispatch command
        fails, nothing else ever frees that spot -- and the lot slowly reports
        itself full while standing empty. Sweeping here (rather than on a
        timer) keeps it lazy and lock-free from the caller's point of view.
        """
        raise NotImplementedError("TODO: reimplement expire_stale_reservations")

    def available_spots(self, car_type: str = "Any") -> list[str]:
        raise NotImplementedError("TODO: reimplement available_spots")

    def reserve_spot(self, spot_name: str, plate: str) -> bool:
        raise NotImplementedError("TODO: reimplement reserve_spot")

    def release_reservation(self, spot_name: str, plate: str) -> None:
        """Undo a reservation when the dispatch command did not actually go out."""
        raise NotImplementedError("TODO: reimplement release_reservation")

    def mark_spot_occupied(self, spot_name: str, plate: str) -> None:
        raise NotImplementedError("TODO: reimplement mark_spot_occupied")

    def mark_spot_vacant(self, spot_name: str) -> None:
        raise NotImplementedError("TODO: reimplement mark_spot_vacant")

    def set_component_broken(self, component_type: str, name: str) -> None:
        raise NotImplementedError("TODO: reimplement set_component_broken")

    def set_component_fixed(self, component_type: str, name: str) -> None:
        raise NotImplementedError("TODO: reimplement set_component_fixed")

    def queue_deferred_repair(self, name: str, component_type: str) -> None:
        raise NotImplementedError("TODO: reimplement queue_deferred_repair")

    def pop_ready_repair(self, name: str) -> Optional[str]:
        """If ``name`` has a pending repair and is now vacant, clear and return its type."""
        raise NotImplementedError("TODO: reimplement pop_ready_repair")

    def update_barrier_state(self, name: str, state_value: str) -> None:
        raise NotImplementedError("TODO: reimplement update_barrier_state")

    def update_zone(self, name: str, co_level: float, danger_level: str) -> None:
        raise NotImplementedError("TODO: reimplement update_zone")

    def co_history(self, zone_name: str) -> list[tuple[float, float]]:
        """Trailing CO readings for a zone, oldest first. Empty if unknown."""
        raise NotImplementedError("TODO: reimplement co_history")

    def occupancy_ratio(self, zone_name: str) -> float:
        """R = (reserved + occupied) / capacity for this zone's parking spots.

        Reserved bays are cars already dispatched but not yet parked - they
        represent load about to land, so counting only OCCUPIED undercounts
        a zone that's about to fill up.
        """
        raise NotImplementedError("TODO: reimplement occupancy_ratio")

    def fans_in_zone(self, zone_name: str) -> list[str]:
        raise NotImplementedError("TODO: reimplement fans_in_zone")

    def lights_in_zone(self, zone_name: str) -> list[str]:
        raise NotImplementedError("TODO: reimplement lights_in_zone")

    def zone_names(self) -> list[str]:
        raise NotImplementedError("TODO: reimplement zone_names")

    # ------------------------------------------------------------------ #
    # Wear tracking (Level 2): cycles on barriers/fans/lights, runtime on
    # fans/lights. Persisted separately via app.db.upsert_component_wear -
    # this only holds the live counters the dashboard reads.
    # ------------------------------------------------------------------ #
    def record_barrier_cycle(self, name: str) -> int:
        """Return movement count, already updated by update_barrier_state()."""
        raise NotImplementedError("TODO: reimplement record_barrier_cycle")

    def set_fan_on(self, name: str, on: bool) -> tuple[int, float]:
        """Toggle a fan and update its wear counters. Returns (cycles, runtime_s)."""
        raise NotImplementedError("TODO: reimplement set_fan_on")

    def set_light_on(self, name: str, on: bool) -> tuple[int, float]:
        raise NotImplementedError("TODO: reimplement set_light_on")

    def wear_snapshot(self) -> list[dict[str, Any]]:
        """Live wear counters for every tracked component, for the wear-threshold sweep."""
        raise NotImplementedError("TODO: reimplement wear_snapshot")

    # ------------------------------------------------------------------ #
    # Vehicle sessions
    # ------------------------------------------------------------------ #
    def start_session(self, plate: str, gate: str, car_type: str = "Normal",
                      planned_minutes: float = 0.0) -> VehicleSession:
        raise NotImplementedError("TODO: reimplement start_session")

    def set_planned_minutes(self, plate: str, planned_minutes: float) -> None:
        """Record the booked duration; later events repeat it, so keep the last."""
        raise NotImplementedError("TODO: reimplement set_planned_minutes")

    def get_session(self, plate: str) -> Optional[VehicleSession]:
        raise NotImplementedError("TODO: reimplement get_session")

    def assign_spot(self, plate: str, spot_name: str) -> None:
        raise NotImplementedError("TODO: reimplement assign_spot")

    def mark_parked(self, plate: str, spot_name: str) -> None:
        raise NotImplementedError("TODO: reimplement mark_parked")

    def mark_left_spot(self, plate: str) -> None:
        """Stop the billing clock when the car vacates its spot."""
        raise NotImplementedError("TODO: reimplement mark_left_spot")

    def mark_exit_requested(self, plate: str, exit_gate: str) -> None:
        raise NotImplementedError("TODO: reimplement mark_exit_requested")

    def mark_at_exit(self, plate: str) -> None:
        raise NotImplementedError("TODO: reimplement mark_at_exit")

    def mark_charged(self, plate: str) -> None:
        raise NotImplementedError("TODO: reimplement mark_charged")

    def mark_paid(self, plate: str, amount: float) -> bool:
        """Returns True iff a session existed, was charged and not already paid."""
        raise NotImplementedError("TODO: reimplement mark_paid")

    def complete_session(self, plate: str) -> Optional[VehicleSession]:
        """Remove the session and hand it back so it can be archived to SQLite."""
        raise NotImplementedError("TODO: reimplement complete_session")

    # ------------------------------------------------------------------ #
    # Telemetry: penalties, activity log, dashboard snapshot
    # ------------------------------------------------------------------ #
    def entry_gates(self) -> list[str]:
        raise NotImplementedError("TODO: reimplement entry_gates")

    def record_penalty(self, reason: str, fine: float, component_type: str, component_name: str) -> None:
        raise NotImplementedError("TODO: reimplement record_penalty")

    def clear_penalties(self) -> None:
        """Forget recorded fines; paired with an admin clearing the penalties table."""
        raise NotImplementedError("TODO: reimplement clear_penalties")

    def reset_for_new_level(self) -> int:
        """Forget the previous level's live park: cars, bays, gates and repair
        bookkeeping. Returns how many vehicle sessions were dropped. Penalties
        and the activity feed are kept - they are the run's record."""
        raise NotImplementedError("TODO: reimplement reset_for_new_level")

    def clear_neglected(self) -> None:
        raise NotImplementedError("TODO: reimplement clear_neglected")

    def log_activity(self, message: str, level: str = "info", capability: str = "logs:view_ops") -> None:
        raise NotImplementedError("TODO: reimplement log_activity")

    def broken_components(self) -> list[dict[str, Any]]:
        raise NotImplementedError("TODO: reimplement broken_components")

    def occupancy_counts(self) -> dict[str, int]:
        raise NotImplementedError("TODO: reimplement occupancy_counts")

    def snapshot(self) -> dict[str, Any]:
        raise NotImplementedError("TODO: reimplement snapshot")

def normalize_car_type(value: str) -> str:
    raise NotImplementedError("TODO: reimplement normalize_car_type")


BarrierGate = Barrier
state = ParkingState(max_processed_events=settings.max_processed_events)
